File: libraries/ary/ary1_vbad.f
      SUBROUTINE ARY1_VBAD( TYPE, N, PNTR, STATUS )
*+
*  Name:
*     ARY1_VBAD

*  Purpose:
*     Set all elements of a vectorised array to the "bad" value.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY1_VBAD( TYPE, N, PNTR, STATUS )

*  Description:
*     The routine sets all elements of a vectorised array, of any
*     numeric data type, to the appropriate "bad" value. The array is
*     passed by pointer.

*  Arguments:
*     TYPE = CHARACTER * ( * ) (Given)
*        An HDS primitive numeric type string, specifying the data type
*        of the vectorised array (case insensitive).
*     N = INTEGER (Given)
*        Number of array elements.
*     PNTR = INTEGER (Given)
*        Pointer to the array whose elements are to be set (the pointer
*        value itself is not altered, although the array elements are).
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Algorithm:
*     -  Test the data type specified against each of the permitted
*     values in turn, calling an appropriate routine to set the array
*     elements to a "bad" value.
*     -  If the data type specified is not valid, then report an error.

*  Copyright:
*     Copyright (C) 1989, 1990 Science & Engineering Research Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     8-JUN-1989  (RFWS):
*        Original version.
*     7-SEP-1989 (RFWS):
*        Added message token for routine name, to prevent '$' from
*        affecting error messages.
*     13-MAR-1990 (RFWS):
*        Changed to call ARY1_BADx instead of VEC_BADx.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'CNF_PAR'          ! For CNV_PVAL function
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_ERR'          ! ARY_ error codes

*  Arguments Given:
      CHARACTER * ( * ) TYPE
      INTEGER N
      INTEGER PNTR

*  Status:
      INTEGER STATUS             ! Global status

*  External references:
      LOGICAL CHR_SIMLR          ! Case insensitive string comparison

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Test the data type string supplied against each permitted value in
*  turn, calling the appropriate routine to set the array elements to a
*  "bad" value.
      IF ( CHR_SIMLR( TYPE, '_BYTE' ) ) THEN
         CALL ARY1_BADB( N, %VAL( CNF_PVAL( PNTR ) ), STATUS )

      ELSE IF ( CHR_SIMLR( TYPE, '_UBYTE' ) ) THEN
         CALL ARY1_BADUB( N, %VAL( CNF_PVAL( PNTR ) ), STATUS )

      ELSE IF ( CHR_SIMLR( TYPE, '_DOUBLE' ) ) THEN
         CALL ARY1_BADD( N, %VAL( CNF_PVAL( PNTR ) ), STATUS )

      ELSE IF ( CHR_SIMLR( TYPE, '_INTEGER' ) ) THEN
         CALL ARY1_BADI( N, %VAL( CNF_PVAL( PNTR ) ), STATUS )

      ELSE IF ( CHR_SIMLR( TYPE, '_REAL' ) ) THEN
         CALL ARY1_BADR( N, %VAL( CNF_PVAL( PNTR ) ), STATUS )

      ELSE IF ( CHR_SIMLR( TYPE, '_WORD' ) ) THEN
         CALL ARY1_BADW( N, %VAL( CNF_PVAL( PNTR ) ), STATUS )

      ELSE IF ( CHR_SIMLR( TYPE, '_UWORD' ) ) THEN
         CALL ARY1_BADUW( N, %VAL( CNF_PVAL( PNTR ) ), STATUS )

*  If the data type string was not valid, then report an error.
      ELSE
         STATUS = ARY__FATIN
         CALL MSG_SETC( 'ROUTINE', 'ARY1_VBAD' )
         CALL MSG_SETC( 'BADTYPE', TYPE )
         CALL ERR_REP( 'ARY1_VBAD_TYPE',
     :   'Routine ^ROUTINE called with an invalid TYPE argument of ' //
     :   '''^BADTYPE'' (internal programming error).', STATUS )
      END IF

*  Call error tracing routine and exit.
      IF ( STATUS .NE. SAI__OK ) CALL ARY1_TRACE( 'ARY1_VBAD', STATUS )

      END

File: libraries/cary/ary1Vbad.c
#include <string.h>
#include "sae_par.h"
#include "ary1.h"
#include "mers.h"
#include "ary_err.h"
#include "dat_par.h"

void ary1Vbad( const char *type, size_t n, void *pntr, int *status ) {
/*
*+
*  Name:
*     ary1Vbad

*  Purpose:
*     Set all elements of a vectorised array to the "bad" value.

*  Synopsis:
*     void ary1Vbad( const char *type, size_t n, void *pntr, int *status )

*  Description:
*     The routine sets all elements of a vectorised array, of any
*     numeric data type, to the appropriate "bad" value. The array is
*     passed by pointer.

*  Parameters:
*     type
*        An HDS primitive numeric type string, specifying the data type
*        of the vectorised array (case insensitive).
*     n
*        Number of array elements.
*     pntr
*        Pointer to the array whose elements are to be set.
*     status
*        The global status.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.

*-
*/

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Test the data type string supplied against each permitted value in turn,
   calling the appropriate routine to set the array elements to a "bad"
   value. */
   if( !strcasecmp( type, "_BYTE" ) ){
      ary1BadB( n, pntr, status );

   } else if( !strcasecmp( type, "_UBYTE" ) ){
      ary1BadUB( n, pntr, status );

   } else if( !strcasecmp( type, "_DOUBLE" ) ){
      ary1BadD( n, pntr, status );

   } else if( !strcasecmp( type, "_INTEGER" ) ){
      ary1BadI( n, pntr, status );

   } else if( !strcasecmp( type, "_REAL" ) ){
      ary1BadF( n, pntr, status );

   } else if( !strcasecmp( type, "_WORD" ) ){
      ary1BadW( n, pntr, status );

   } else if( !strcasecmp( type, "_UWORD" ) ){
      ary1BadUW( n, pntr, status );

   } else if( !strcasecmp( type, "_INT64" ) ){
      ary1BadK( n, pntr, status );

/* If the data type string was not valid, then report an error. */
   } else {
      *status = ARY__FATIN;
      msgSetc( "BADTYPE", type );
      errRep( "ARY1_VBAD_TYPE",
              "Routine ary1Vbad called with an invalid 'type' argument of"
              "'^BADTYPE' (internal programming error).", status );
   }

/* Call error tracing routine and exit. */
   if( *status != SAI__OK ) ary1Trace( "ary1Vbad", status );

}
